### DISCUSSION TITLE
Introducing convenience macros to shadow's config file?

### DISCUSSION THREAD

**Discussion by rwails** [shadow/shadow#2618] [Ideas]
I think it's fairly common in Shadow workflows to want to specify input files for the managed processes in the "experiment" root directory. The way to do this with relative paths is pretty clunky at the moment: e.g., in the tgen example in the getting started guide, here's the recommended way 

args: ../../../tgen.server.graphml.xml

<https://shadow.github.io/docs/guide/getting_started_tgen.html>

I wonder if some convenience macros could make this more user friendly. Maybe variables like:

$CWD  <-> current working directory that Shadow is invoked from
$DATA_PARENT / $EXPERIMENT_DIR <-> the parent of the shadow data directory; roughly speaking, the "experiment" directory.

Then the above example would simplify to:

args: $EXPERIMENT_DIR/tgen.server.graphml.xml

wdyt?

**Comment by sporksmith** [comment]
We keep talking about ideas like this, but have so far decided to steer away from the rabbit hole of evolving an ad-hoc configuration/scripting language, in favor of recommending users to generate configs using some other language for most non-trivial configs. I recently wrote up some small examples here: https://shadow.github.io/docs/guide/shadow_config_complex.html

It might be nice to eventually be more opinionated though and pick a language to use in our documentation, maybe provide some modules/libraries for that language, or maybe even embed an interpreter in Shadow.

**Comment by stevenengler** [comment]
If we want to keep the config parsing simple, maybe we could expose these things to the process using environment variables. So once we support fork/exec, you could do something like:

```yaml
path: bash
args: -c 'tgen $EXPERIMENT_DIR/tgen.server.graphml.xml'
```

But maybe this is even more clunky than the original method.

**Comment by robgjansen** [comment]
I think we've all found it slightly annoying at times. However, I believe that the maintenance burden of supporting ad-hoc variables or a special interpreter for the config files currently outweighs the incremental usability benefit that would be experienced (relative to [the extended yaml examples from our guide](https://shadow.github.io/docs/guide/shadow_config_complex.html)).
